docs/reinterpreting_effects.md

```diff
@@ -0,0 +1,359 @@
+
+## "Stacking" effects
+
+### The production use-case
+
+Now that we have these two mini-DSL effect types established, we need to stitch them together.
+
+Let's take a moment to think about what could go wrong with an HTTP API from which we plan to fetch some JSON and convert it into a list of `CatFact`s.
+
+We can conceive that the server might occasionally return a malformed JSON response:
+
+``` haskell
+newtype JsonParseError = JsonParseError String
+  deriving (Show, Eq)
+
+decodeOrThrow :: (Has (Throw JsonParseError) sig m, FromJSON a) => L.ByteString -> m a
+decodeOrThrow = either (throwError . JsonParseError) pure . eitherDecode
+```
+
+A more HTTP-centric issue is that we might receive a content type we can't use. In this case, anything that's not `application/json`:
+
+``` haskell
+newtype InvalidContentType = InvalidContentType String
+  deriving (Show, Eq)
+
+```
+
+Now we need to support fetching JSON given an HTTP request. We have no guarantee that an arbitrary HTTP request
+will actually return JSON, so for this implementation we have to account for failure conditions. This provides
+a great opportunity to see how effect handlers can actually rely on *multiple underlying effects*!
+
+``` haskell
+newtype CatFactsApi m a = CatFactsApi { runCatFactsApi :: m a }
+ deriving newtype
+      ( Monad
+      , Functor
+      , Applicative
+      , MonadIO
+      , Alternative
+      )
+
+catFactsEndpoint :: HTTP.Request
+catFactsEndpoint = HTTP.parseRequest_ "https://cat-fact.herokuapp.com/facts/random"
+
+instance ( Has Http sig m
+         , Has (Throw JsonParseError) sig m
+         , Has (Throw InvalidContentType) sig m
+         , Algebra sig m
+         ) =>
+         Algebra (CatFactClient :+: sig) (CatFactsApi m) where
+  eff (L (ListFacts numberOfFacts k)) = do
+    resp <- sendRequest (catFactsEndpoint { HTTP.queryString = "?amount=" <> B.pack (show numberOfFacts) })
+    case lookup hContentType (HTTP.responseHeaders resp) of
+      Just "application/json; charset=utf-8" -> decodeOrThrow (HTTP.responseBody resp) >>= k
+      other -> throwError (InvalidContentType (show other))
+  eff (R other) = CatFactsApi (eff (handleCoercible other))
+```
+
+We implement a `CatFacts` effect handler that depends on _three_ underlying effects:
+
+1. `Http` - we need to be able to make requests
+2. `Throw JsonParseError` - we need to be able to signal that some aspect of the JSON wasn't what we expected.
+3. `Throw InvalidContentType` - we need to be able to signal what we got wasn't JSON at all!
+
+The nice aspect of this is that we have neatly contained the failure scenarios to their relevant strata rather than leaking them into the higher-level abstraction (`listFacts`)!
```

Comment:
This is really lovely!

There’s another possible implementation that might be preferable in some situations, which is to define the `CatFactsApi` carrier as a wrapper for some or all of the required carriers, e.g.:

```haskell
newtype CatFactsApi m a = CatFactsApi { runCatFactsApi :: ThrowC JsonParseError (ThrowC InvalidContentType (HttpClient m)) a }
```

If we do this, we don’t need the `Has` constraints, but on the other hand we might be making redundant work for ourselves if other stuff we use can also throw those errors, and we’d absolutely be limiting our ability to mock out the http client. This might be worth mentioning in passing, because the tradeoffs are super interesting and relevant when designing carriers; but I can also see an argument for pushing that to a future PR/document. What do you think?

Reply:
Hm, I feel like this might merit its own post. I'm not convinced that forcing the implementation to exclusively use `ThrowC` as the only carrier for `Throw` in this case is ideal. This document is focused on promoting effect reinterpretation, so this seems to run counter to that point and might cause confusion.

It might be the case though that I'm missing some benefits of the tradeoff beyond eliding some `Has` constraints?

Reply:
@iand675 abstraction, basically—you can completely encapsulate the underlying action. Think of `Control.Carrier.Writer.Strict.WriterC`, which is implemented as a wrapper around `StateC`—if it were implemented using a `Has` constraint instead, then it would be easy for another action to `put mempty`, violating `Writer`’s invariants. Reinterpreting-by-wrapping allows us to use `StateC`—and even the `State` operations—without having to expose the `State w` effect to the rest of the program.

All of which is to say, I think you’re exactly right that this deserves its own post 👍